src/java/main/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -303,15 +305,38 @@ public void loadData() throws IOException, InterruptedException {
 
     public void takeSnapshot(){
```

Comment:
@flier Having SyncRequestProcessor look at the number of snapshot files feels like a bit of a hack, but I am not expert here.  I would be fine with having SyncRequestProcessor delegate all throttling of snapshots to ZooKeeperServer, but then we need some sort of synchronization to prevent a snapshot from being taken when it is not in a proper state (during the ZABP sync phase).